### PR TITLE
Fixing infinite loop/crash when no result sets have a pinPercentage

### PR DESCRIPTION
**Pull request by ryanml**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -370,6 +370,10 @@ Synopsis.prototype.winners = function (n, weights) {
 
   if (!results) return
 
+  // If none of the results have a defined pinPercentage, return the empty winners list
+  const noPinPercentage = results.filter(r => typeof r.pinPercentage === 'undefined')
```
Comment (NejcZdovc) [inline index.js]: What happens if pinPercentage is `null`

Comment (ryanml) [reply index.js]: @NejcZdovc can check for null as well, in the crash that was being investigated it seemed that this only was ever undefined.

**Review by NejcZdovc — DISMISSED** [review]
Please add code coverage for this function

**Review by mrose17 — DISMISSED** [review]
LGTM

File: index.js [hunk 2]
```diff
@@ -362,14 +362,27 @@ Synopsis.prototype.winner = function () {
...
   const pinned = []
   const winners = []
   let count = 0
   let allP = true
 
   if (!results) return
 
+  results = results.filter((result) => {
+    if (
+      result.pinPercentage == null ||
```
Comment (NejcZdovc) [inline index.js]: Isn't `result.pinPercentage == null` enough? it will check null and undefined

File: index.js [hunk 3]
```diff
@@ -362,14 +362,24 @@ Synopsis.prototype.winner = function () {
...
   const pinned = []
   const winners = []
   let count = 0
   let allP = true
 
   if (!results) return
 
+  results = results.filter((result) => {
+    if (result.pinPercentage == null) {
+      return result.weight !== 0
```
Comment (NejcZdovc) [inline index.js]: weight should be > 0

**Review by NejcZdovc — COMMENTED** [review]
Looks good to me, let's wait for @bsclifton to confirm that this fixes the problem with the broken profile

File: test/winners.js [hunk 4]
```diff
@@ -25,4 +25,58 @@ weights = [
   { publisher: '17%', weight: 0.17 }
 ]
 tap.same(_.sortBy(Synopsis.prototype.winners(10, weights)),
-         [ '05%', '10%', '17%', '17%', '17%', '20%', '20%', '30%', '30%', '30%' ])
+  [ '05%', '10%', '17%', '17%', '17%', '20%', '20%', '30%', '30%', '30%' ])
+
+tap.same([], Synopsis.prototype.winners(92,
+  [
+    { publisher: 'brave.com', pinPercentage: undefined, weight: 0 },
+    { publisher: 'clifton.io', pinPercentage: undefined, weight: 0 }
```
Comment (bsclifton) [inline test/winners.js:39]: can we remove the 2nd item here? so that it matches the test case we had originally

Comment (bsclifton) [reply test/winners.js:39]: (see https://github.com/bsclifton/bat-publisher/commit/d3364cf17aa0c31922c21b17f972b936c92e45ee)

**Review by NejcZdovc — APPROVED** [review]
++

File: test/winners.js (hunk 4 above)
Comment (ryanml) [reply test/winners.js:39]: @bsclifton additional test added here: https://github.com/brave-intl/bat-publisher/pull/35/files#diff-62377b611eef58dc8064a39a7b75112dR30

Only difference from original test case is that we are returning `[]` as opposed to `undefined` (winners should be an array in all cases)

Comment (bsclifton) [reply test/winners.js:39]: yup! that is perfect

**Review by bsclifton — APPROVED** [review]
😄 👌

**Pull request merged**